File: sidecar/src/config/import-parser.ts
import { parse as parseJsonc, printParseErrorCode, type ParseError } from "jsonc-parser";
import { parse as parseToml } from "smol-toml";
import type { ScannedServer, UnsupportedServer, ImportDiagnostic, ParsedImport } from "@moor/types";

export type { ScannedServer, UnsupportedServer, ImportDiagnostic, ParsedImport };

const HTTP_TYPES = new Set(["http", "sse", "streamable-http", "streamable_http", "remote"]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value : undefined;
}

function asStringArray(value: unknown): string[] | undefined {
  if (!Array.isArray(value)) return undefined;
  const values = value.filter((item): item is string => typeof item === "string");
  return values.length === value.length ? values : undefined;
}

function asStringRecord(value: unknown): Record<string, string> | undefined {
  if (!isRecord(value)) return undefined;
  const entries = Object.entries(value)
    .map(([key, item]) => {
      if (typeof item === "string") return [key, item] as const;
      if (typeof item === "number" || typeof item === "boolean") {
        return [key, String(item)] as const;
      }
      return null;
    })
    .filter((entry): entry is readonly [string, string] => entry !== null);
  return entries.length > 0 ? Object.fromEntries(entries) : undefined;
}

function asHeaderRecord(value: unknown): Record<string, string> | undefined {
  if (!isRecord(value)) return undefined;
  const entries = Object.entries(value)
    .map(([key, item]) => {
      if (typeof item === "string") return [key, item] as const;
      if (isRecord(item) && typeof item.env === "string") {
        return [key, `{env:${item.env}}`] as const;
      }
      return null;
    })
    .filter((entry): entry is readonly [string, string] => entry !== null);
  return entries.length > 0 ? Object.fromEntries(entries) : undefined;
}

function envHeaders(value: unknown): Record<string, string> | undefined {
  if (!isRecord(value)) return undefined;
  const entries = Object.entries(value)
    .map(([key, item]): [string, string] | null =>
      typeof item === "string" ? [key, `{env:${item}}`] : null,
    )
    .filter((entry): entry is [string, string] => entry !== null);
  return entries.length > 0 ? Object.fromEntries(entries) : undefined;
}

function mergeRecords(
  ...records: Array<Record<string, string> | undefined>
): Record<string, string> | undefined {
  const merged: Record<string, string> = Object.assign({}, ...records.filter(Boolean));
  return Object.keys(merged).length > 0 ? merged : undefined;
}

function normalizeServer(
  name: string,
  rawConfig: Record<string, unknown>,
  source: string,
): ScannedServer | UnsupportedServer | null {
  if (rawConfig.enabled === false) return null;

  const type = asString(rawConfig.type)?.toLowerCase();
  if (type === "openapi" || rawConfig.openapi) {
    return { name, source, reason: "OpenAPI-to-MCP is not supported" };
  }

  const commandArray = asStringArray(rawConfig.command);
  const command = asString(rawConfig.command) ?? commandArray?.[0];
  const args = commandArray?.slice(1) ?? asStringArray(rawConfig.args);
  const url = asString(rawConfig.url);
  const env = asStringRecord(rawConfig.env) ?? asStringRecord(rawConfig.environment);
  const workingDir =
    asString(rawConfig.cwd) ?? asString(rawConfig.workingDir) ?? asString(rawConfig.working_dir);
  const bearerTokenEnv = asString(rawConfig.bearer_token_env_var);
  const headers = mergeRecords(
    asHeaderRecord(rawConfig.headers),
    asHeaderRecord(rawConfig.http_headers),
    envHeaders(rawConfig.env_http_headers),
    bearerTokenEnv ? { Authorization: `Bearer {env:${bearerTokenEnv}}` } : undefined,
  );

  if (type === "local" || type === "stdio" || command) {
    if (!command) return { name, source, reason: "stdio server is missing a command" };
    return {
      name,
      connectionType: "stdio",
      command,
      args,
      env,
      workingDir,
      source,
    };
  }

  if (url && (!type || HTTP_TYPES.has(type))) {
    return {
      name,
      connectionType: "http",
      url,
      headers,
      source,
    };
  }

  if (type && !HTTP_TYPES.has(type)) {
    return { name, source, reason: `unsupported server type "${type}"` };
  }

  return { name, source, reason: "config is missing command or url" };
}

function parseServerMap(value: unknown, source: string): ParsedImport {
  const servers: ScannedServer[] = [];
  const unsupported: UnsupportedServer[] = [];

  if (!isRecord(value)) {
    return {
      servers,
      unsupported,
      errors: [`${source}: no valid server map found`],
      diagnostics: [],
    };
  }

  for (const [name, rawConfig] of Object.entries(value)) {
    if (!isRecord(rawConfig)) {
      unsupported.push({ name, source, reason: "server config must be an object" });
      continue;
    }

    const normalized = normalizeServer(name, rawConfig, source);
    if (!normalized) continue;
    if ("connectionType" in normalized) servers.push(normalized);
    else unsupported.push(normalized);
  }

  return { servers, unsupported, errors: [], diagnostics: [] };
}

function mergeParsed(...results: ParsedImport[]): ParsedImport {
  return {
    servers: results.flatMap((result) => result.servers),
    unsupported: results.flatMap((result) => result.unsupported),
    errors: results.flatMap((result) => result.errors),
    diagnostics: results.flatMap((result) => result.diagnostics),
  };
}

function lineColumnAtOffset(content: string, offset: number): { line: number; column: number } {
  const previousLineBreak = content.lastIndexOf("\n", Math.max(0, offset - 1));
  return {
    line: content.slice(0, offset).split("\n").length,
    column: offset - previousLineBreak,
  };
}

function toJsonDiagnostic(source: string, content: string, error: ParseError): ImportDiagnostic {
  const position = lineColumnAtOffset(content, error.offset);
  const code = printParseErrorCode(error.error);
  return {
    source,
    message: code,
    code,
    line: position.line,
    column: position.column,
    offset: error.offset,
    length: error.length,
  };
}

export function parseJsonMcpConfig(content: string, source: string): ParsedImport {
  const errors: ParseError[] = [];
  const config = parseJsonc(content, errors, { allowTrailingComma: true }) as unknown;
  if (errors.length > 0) {
    const diagnostics = errors.map((error) => toJsonDiagnostic(source, content, error));
    const first = diagnostics[0];
    return {
      servers: [],
      unsupported: [],
      errors: [
        first?.line && first.column
          ? `${source}: JSON parse error at line ${first.line}, column ${first.column}`
          : `${source}: JSON parse error`,
      ],
      diagnostics,
    };
  }
  if (!isRecord(config)) {
    return {
      servers: [],
      unsupported: [],
      errors: [`${source}: config root must be an object`],
      diagnostics: [],
    };
  }

  const results: ParsedImport[] = [];
  if ("mcpServers" in config) results.push(parseServerMap(config.mcpServers, source));
  if ("mcp" in config) results.push(parseServerMap(config.mcp, source));

  return results.length > 0
    ? mergeParsed(...results)
    : {
        servers: [],
        unsupported: [],
        errors: [`${source}: no mcpServers or mcp key found`],
        diagnostics: [],
      };
}

export function parseCodexTomlConfig(content: string, source: string): ParsedImport {
  try {
    const config = parseToml(content) as unknown;
    if (!isRecord(config)) {
      return {
        servers: [],
        unsupported: [],
        errors: [`${source}: TOML root must be an object`],
        diagnostics: [],
      };
    }
    if (!("mcp_servers" in config)) {
      return {
        servers: [],
        unsupported: [],
        errors: [`${source}: no mcp_servers key found`],
        diagnostics: [],
      };
    }
    return parseServerMap(config.mcp_servers, source);
  } catch {
    return {
      servers: [],
      unsupported: [],
      errors: [`${source}: TOML parse error`],
      diagnostics: [],
    };
  }
}

File: src-tauri/src/sidecar/config/import_parser.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScannedServer {
    pub name: String,
    pub connection_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub args: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env: Option<std::collections::HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub headers: Option<std::collections::HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_dir: Option<String>,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnsupportedServer {
    pub name: String,
    pub source: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportDiagnostic {
    pub source: String,
    pub message: String,
    pub code: Option<String>,
    pub line: Option<usize>,
    pub column: Option<usize>,
    pub offset: Option<usize>,
    pub length: Option<usize>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedImport {
    pub servers: Vec<ScannedServer>,
    pub unsupported: Vec<UnsupportedServer>,
    pub errors: Vec<String>,
    pub diagnostics: Vec<ImportDiagnostic>,
}

const HTTP_TYPES: &[&str] = &[
    "http",
    "sse",
    "streamable-http",
    "streamable_http",
    "remote",
];

pub fn parse_json_mcp_config(content: &str, source: &str) -> ParsedImport {
    let stripped = strip_jsonc(content);
    let config: Value = match serde_json::from_str(&stripped) {
        Ok(v) => v,
        Err(e) => {
            let line = e.line();
            let column = e.column();
            return ParsedImport {
                servers: vec![],
                unsupported: vec![],
                errors: vec![if line > 0 && column > 0 {
                    format!("{source}: JSON parse error at line {line}, column {column}")
                } else {
                    format!("{source}: JSON parse error")
                }],
                diagnostics: vec![ImportDiagnostic {
                    source: source.to_string(),
                    message: e.to_string(),
                    code: None,
                    line: Some(line),
                    column: Some(column),
                    offset: None,
                    length: None,
                }],
            };
        }
    };

    let config_obj = match config.as_object() {
        Some(obj) => obj,
        None => {
            return ParsedImport {
                errors: vec![format!("{source}: config root must be an object")],
                ..Default::default()
            }
        }
    };

    let mut results = Vec::new();
    if let Some(mcp_servers) = config_obj.get("mcpServers") {
        results.push(parse_server_map(mcp_servers, source));
    }
    if let Some(mcp) = config_obj.get("mcp") {
        results.push(parse_server_map(mcp, source));
    }

    if results.is_empty() {
        return ParsedImport {
            errors: vec![format!("{source}: no mcpServers or mcp key found")],
            ..Default::default()
        };
    }

    merge_parsed(&results)
}

pub fn parse_codex_toml_config(content: &str, source: &str) -> ParsedImport {
    let config: toml::Value = match toml::from_str(content) {
        Ok(v) => v,
        Err(_) => {
            return ParsedImport {
                errors: vec![format!("{source}: TOML parse error")],
                ..Default::default()
            }
        }
    };

    let config_table = match config.as_table() {
        Some(t) => t,
        None => {
            return ParsedImport {
                errors: vec![format!("{source}: TOML root must be an object")],
                ..Default::default()
            }
        }
    };

    match config_table.get("mcp_servers") {
        Some(mcp_servers) => {
            let json_val = toml_to_json_value(mcp_servers);
            parse_server_map(&json_val, source)
        }
        None => ParsedImport {
            errors: vec![format!("{source}: no mcp_servers key found")],
            ..Default::default()
        },
    }
}

fn toml_to_json_value(val: &toml::Value) -> serde_json::Value {
    match val {
        toml::Value::String(s) => serde_json::Value::String(s.clone()),
        toml::Value::Integer(i) => serde_json::json!(*i),
        toml::Value::Float(f) => serde_json::json!(*f),
        toml::Value::Boolean(b) => serde_json::json!(*b),
        toml::Value::Array(arr) => {
            serde_json::Value::Array(arr.iter().map(toml_to_json_value).collect())
        }
        toml::Value::Table(tbl) => {
            let mut map = serde_json::Map::new();
            for (k, v) in tbl {
                map.insert(k.clone(), toml_to_json_value(v));
            }
            serde_json::Value::Object(map)
        }
        toml::Value::Datetime(dt) => serde_json::Value::String(dt.to_string()),
    }
}

fn parse_server_map(value: &Value, source: &str) -> ParsedImport {
    let obj = match value.as_object() {
        Some(o) => o,
        None => {
            return ParsedImport {
                errors: vec![format!("{source}: no valid server map found")],
                ..Default::default()
            }
        }
    };

    let mut servers = vec![];
    let mut unsupported = vec![];

    for (name, raw_config) in obj {
        let raw_obj = match raw_config.as_object() {
            Some(o) => o,
            None => {
                unsupported.push(UnsupportedServer {
                    name: name.clone(),
                    source: source.to_string(),
                    reason: "server config must be an object".to_string(),
                });
                continue;
            }
        };

        if raw_obj.get("enabled").and_then(|v| v.as_bool()) == Some(false) {
            continue;
        }

        match normalize_server(name, raw_obj, source) {
            Some(Normalized::Server(s)) => servers.push(s),
            Some(Normalized::Unsupported(u)) => unsupported.push(u),
            None => {}
        }
    }

    ParsedImport {
        servers,
        unsupported,
        errors: vec![],
        diagnostics: vec![],
    }
}

enum Normalized {
    Server(ScannedServer),
    Unsupported(UnsupportedServer),
}

fn normalize_server(
    name: &str,
    raw: &serde_json::Map<String, Value>,
    source: &str,
) -> Option<Normalized> {
    let type_val = raw
        .get("type")
        .and_then(|v| v.as_str())
        .map(|s| s.to_lowercase());

    if type_val.as_deref() == Some("openapi") || raw.contains_key("openapi") {
        return Some(Normalized::Unsupported(UnsupportedServer {
            name: name.to_string(),
            source: source.to_string(),
            reason: "OpenAPI-to-MCP is not supported".to_string(),
        }));
    }

    let command_array = raw.get("command").and_then(|v| v.as_array());
    let command = raw
        .get("command")
        .and_then(|v| v.as_str())
        .map(String::from)
        .or_else(|| {
            command_array.and_then(|arr| arr.first().and_then(|v| v.as_str()).map(String::from))
        });
    let args: Option<Vec<String>> = command_array
        .map(|arr| {
            arr.iter()
                .skip(1)
                .filter_map(|v| v.as_str().map(String::from))
                .collect::<Vec<String>>()
        })
        .filter(|a: &Vec<String>| !a.is_empty())
        .or_else(|| {
            raw.get("args").and_then(|v| v.as_array()).map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect::<Vec<String>>()
            })
        });
    let url = raw.get("url").and_then(|v| v.as_str()).map(String::from);
    let env = as_string_record(raw.get("env").or(raw.get("environment")));
    let working_dir = raw
        .get("cwd")
        .or(raw.get("workingDir"))
        .or(raw.get("working_dir"))
        .and_then(|v| v.as_str())
        .map(String::from);
    let bearer_token_env = raw.get("bearer_token_env_var").and_then(|v| v.as_str());
    let headers = merge_records(
        as_header_record(raw.get("headers")),
        as_header_record(raw.get("http_headers")),
        env_headers(raw.get("env_http_headers")),
        bearer_token_env.map(|var| {
            let mut map = std::collections::HashMap::new();
            map.insert("Authorization".to_string(), format!("Bearer {{env:{var}}}"));
            map
        }),
    );

    let is_local = type_val.as_deref() == Some("local") || type_val.as_deref() == Some("stdio");
    if is_local || command.is_some() {
        let command = match command {
            Some(c) => c,
            None => {
                return Some(Normalized::Unsupported(UnsupportedServer {
                    name: name.to_string(),
                    source: source.to_string(),
                    reason: "stdio server is missing a command".to_string(),
                }))
            }
        };
        return Some(Normalized::Server(ScannedServer {
            name: name.to_string(),
            connection_type: "stdio".to_string(),
            command: Some(command),
            args,
            url: None,
            env,
            headers: None,
            working_dir,
            source: source.to_string(),
        }));
    }

    if let Some(url) = url {
        if type_val
            .as_deref()
            .map(|t| HTTP_TYPES.contains(&t))
            .unwrap_or(true)
        {
            return Some(Normalized::Server(ScannedServer {
                name: name.to_string(),
                connection_type: "http".to_string(),
                command: None,
                args: None,
                url: Some(url),
                env: None,
                headers,
                working_dir: None,
                source: source.to_string(),
            }));
        }
        return Some(Normalized::Unsupported(UnsupportedServer {
            name: name.to_string(),
            source: source.to_string(),
            reason: format!(
                "unsupported server type \"{}\"",
                type_val.unwrap_or_default()
            ),
        }));
    }

    if let Some(ref t) = type_val {
        if !HTTP_TYPES.contains(&t.as_str()) {
            return Some(Normalized::Unsupported(UnsupportedServer {
                name: name.to_string(),
                source: source.to_string(),
                reason: format!("unsupported server type \"{t}\""),
            }));
        }
    }

    Some(Normalized::Unsupported(UnsupportedServer {
        name: name.to_string(),
        source: source.to_string(),
        reason: "config is missing command or url".to_string(),
    }))
}

fn as_string_record(value: Option<&Value>) -> Option<std::collections::HashMap<String, String>> {
    let obj = value?.as_object()?;
    let mut map = std::collections::HashMap::new();
    for (k, v) in obj {
        match v {
            Value::String(s) => {
                map.insert(k.clone(), s.clone());
            }
            Value::Number(n) => {
                map.insert(k.clone(), n.to_string());
            }
            Value::Bool(b) => {
                map.insert(k.clone(), b.to_string());
            }
            _ => continue,
        }
    }
    if map.is_empty() {
        None
    } else {
        Some(map)
    }
}

fn as_header_record(value: Option<&Value>) -> Option<std::collections::HashMap<String, String>> {
    let obj = value?.as_object()?;
    let mut map = std::collections::HashMap::new();
    for (k, v) in obj {
        match v {
            Value::String(s) => {
                map.insert(k.clone(), s.clone());
            }
            Value::Object(inner) => {
                if let Some(env_val) = inner.get("env").and_then(|v| v.as_str()) {
                    map.insert(k.clone(), format!("{{env:{env_val}}}"));
                }
            }
            _ => continue,
        }
    }
    if map.is_empty() {
        None
    } else {
        Some(map)
    }
}

fn env_headers(value: Option<&Value>) -> Option<std::collections::HashMap<String, String>> {
    let obj = value?.as_object()?;
    let mut map = std::collections::HashMap::new();
    for (k, v) in obj {
        if let Value::String(s) = v {
            map.insert(k.clone(), format!("{{env:{s}}}"));
        }
    }
    if map.is_empty() {
        None
    } else {
        Some(map)
    }
}

fn merge_records(
    a: Option<std::collections::HashMap<String, String>>,
    b: Option<std::collections::HashMap<String, String>>,
    c: Option<std::collections::HashMap<String, String>>,
    d: Option<std::collections::HashMap<String, String>>,
) -> Option<std::collections::HashMap<String, String>> {
    let mut merged = std::collections::HashMap::new();
    for map in [a, b, c, d].into_iter().flatten() {
        merged.extend(map);
    }
    if merged.is_empty() {
        None
    } else {
        Some(merged)
    }
}

fn merge_parsed(results: &[ParsedImport]) -> ParsedImport {
    let mut servers = vec![];
    let mut unsupported = vec![];
    let mut errors = vec![];
    let mut diagnostics = vec![];
    for r in results {
        servers.extend(r.servers.clone());
        unsupported.extend(r.unsupported.clone());
        errors.extend(r.errors.clone());
        diagnostics.extend(r.diagnostics.clone());
    }
    ParsedImport {
        servers,
        unsupported,
        errors,
        diagnostics,
    }
}

/// Strip JSONC comments (// and /* */) and trailing commas from JSON content.
fn strip_jsonc(content: &str) -> String {
    let mut result = String::with_capacity(content.len());
    let chars: Vec<char> = content.chars().collect();
    let len = chars.len();
    let mut i = 0;
    let mut in_string = false;

    while i < len {
        let c = chars[i];

        if in_string {
            result.push(c);
            if c == '\\' && i + 1 < len {
                i += 1;
                result.push(chars[i]);
            } else if c == '"' {
                in_string = false;
            }
            i += 1;
            continue;
        }

        match c {
            '"' => {
                in_string = true;
                result.push(c);
            }
            '/' if i + 1 < len && chars[i + 1] == '/' => {
                while i < len && chars[i] != '\n' {
                    i += 1;
                }
                continue;
            }
            '/' if i + 1 < len && chars[i + 1] == '*' => {
                i += 2;
                while i + 1 < len && !(chars[i] == '*' && chars[i + 1] == '/') {
                    i += 1;
                }
                i += 2;
                continue;
            }
            ',' if i + 1 < len => {
                let mut j = i + 1;
                while j < len && chars[j].is_whitespace() {
                    j += 1;
                }
                if j < len && (chars[j] == ']' || chars[j] == '}') {
                    i += 1;
                    continue;
                }
                result.push(c);
            }
            _ => {
                result.push(c);
            }
        }
        i += 1;
    }
    result
}
